### PR TITLE
Directo: Reference BL from gem

### LINKED ISSUE #509: Refactor models/directo.rb
- Consider extracting Directo into separate gem
- Enable accounting software easy switch
- AccountActivity > AccountEntry

#205 

API spec
http://wiki.directo.ee/et/xmlcore

### PR DESCRIPTION
**Pull request by karlerikounapuu**
Closes #509 

Sending monthly registrar purchase summary:
`DirectoInvoiceForwardJob.run(monthly: true)`

Sending (daily / whenever) registrar prepayment invoices:
`DirectoInvoiceForwardJob.run`

Appending `dry: true` to `run` arguments composes invoice list, but doesn't deliver invoices to API.

Ready for testing, but please do not merge yet as it references development branch of Directo gem.

### CODE REVIEW

File: app/models/concerns/invoice/book_keeping.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module Concerns
+  module Invoice
+    module BookKeeping
+      extend ActiveSupport::Concern
+
+      def as_directo_json
+        inv = ActiveSupport::JSON.decode(ActiveSupport::JSON.encode(self))
```
Comment (yulgolem) [inline app/models/concerns/invoice/book_keeping.rb]: Maybe 'invoice'? Just 'inv' seems not descriptive enough.

**Comment by vohmar** [comment]
there is no feedback in stdout

both jobs fail. sample error from logs:

> Mar  3 15:23:38 staging st-adm.infra.tld.ee[23283]: [Directo] - attempting to send following XML:#012 <?xml version="1.0" encoding="UTF-8"?><invoices>  <invoice Number="309910" InvoiceDate="2020-02-29" PaymentTerm="R" CustomerCode="EEDIRECT" CustomerName="" Language="ENG" Currency="EUR" SalesAgent="JAANA" TotalVAT="" TotalWoVAT="">    <line RN="1" RR="1" ProductName="Domain registrations - February 2020"/>    <line RN="2" RR="2" ProductID="01EE" Quantity="7" Unit="pc" ProductName=".ee registration: 1 year(s)" UnitPriceWoVAT="6.0"/>    <line RN="3" RR="3" ProductID="ETTEM06" Quantity="1" Unit="pc" ProductName="Domains prepayment" UnitPriceWoVAT="42.0"/>  </invoice>  <invoice Number="309911" InvoiceDate="2020-02-29" PaymentTerm="R" CustomerCode="RIKS" CustomerName="" Language="" Currency="EUR" SalesAgent="JAANA" TotalVAT="" TotalWoVAT="">    <line RN="1" RR="1" ProductName="Domeenide registreerimine - Veebruar 2020"/>    <line RN="2" RR="2" ProductID="01EE" Quantity="5" Unit="tk" ProductName=".ee registreerimine: 1 aasta(t)" UnitPriceWoVAT="6.0"/>    <line RN="3" RR="3" ProductID="01EE" Quantity="2" Unit="tk" ProductName=".ee registreerimine: 3 kuu(d)" UnitPriceWoVAT="1.6"/>    <line RN="4" RR="4" ProductID="ETTEM06" Quantity="1" Unit="tk" ProductName="Domeenide ettemaks" UnitPriceWoVAT="33.2"/>  </invoice></invoices>
> Mar  3 15:23:38 staging st-adm.infra.tld.ee[23283]: [Directo] - Responded with body: <?xml version="1.0" encoding="utf-8"?><error code="20" desc="invalid xml given" err_reason="Invalid character value for cast specification." />

**Comment by vohmar** [comment]
still no feedback in stdout

sending monthly inv succeeded
sending regular invoices failed with no error:

```
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: [DIRECTO] Trying to send 2 prepayment invoices
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: Invoice Load (1.2ms)  SELECT "invoices".* FROM "invoices" WHERE "inv
oices"."in_directo" = $1 AND "invoices"."cancelled_at" IS NULL  [["in_directo", "f"]]
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: AccountActivity Load (0.8ms)  SELECT  "account_activities".* FROM "a
ccount_activities" WHERE "account_activities"."invoice_id" = $1 LIMIT $2  [["invoice_id", 94], ["LIMIT", 1]]
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: [DIRECTO] Invoice 3131076 has been skipped
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: AccountActivity Load (0.4ms)  SELECT  "account_activities".* FROM "a
ccount_activities" WHERE "account_activities"."invoice_id" = $1 LIMIT $2  [["invoice_id", 105], ["LIMIT", 1]]
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: [DIRECTO] Invoice 3131087 has been skipped
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: [Directo] - attempting to send following XML:#012 <?xml version="1.0
" encoding="UTF-8"?><invoices/>
Mar 10 15:17:02 staging st-adm.infra.tld.ee[26381]: [Directo] - Responded with body:
```

for some reason sending the first invoice was skipped. Second one is not paid so this is expected. First one has record in account_activities table and looks ok - what could be the reason here?

**Comment by vohmar** [comment]
testing done, did not find any issues ready to merge

**Pull request merged**